[tensorexpr] Add python bindings for TensorExprKernel

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54450 [tensorexpr] Add python bindings for TensorExprKernel**

Differential Revision: [D27243175](https://our.internmc.facebook.com/intern/diff/D27243175)

huiguoo committed:
[tensorexpr] Add python bindings for TensorExprKernel

[ghstack-poisoned]

huiguoo commented:
@ZolotukhinM I didn't try hard to test this in python. It seems not easy to extract the correct subgraph and prepare a stack from scratch. I'll skip it now in case you already have a test example :)

ZolotukhinM commented:
> It seems not easy to extract the correct subgraph and prepare a stack from scratch

It's actually very important to make this part easy, since that's the main function of the class - without it it's hard to use the new bindings anyway.

You could probably do something similar to this to construct a `Stack`:
https://github.com/pytorch/pytorch/blob/0f628d1503503bc9c92709376c6c1a4f73b80052/torch/csrc/jit/python/init.cpp#L420-L436

huiguoo committed:
Update on "[tensorexpr] Add python bindings for TensorExprKernel"


Differential Revision: [D27243175](https://our.internmc.facebook.com/intern/diff/D27243175)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add python bindings for TensorExprKernel"


Differential Revision: [D27243175](https://our.internmc.facebook.com/intern/diff/D27243175)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add python bindings for TensorExprKernel"


Differential Revision: [D27243175](https://our.internmc.facebook.com/intern/diff/D27243175)

[ghstack-poisoned]

huiguoo committed:
Update on "[tensorexpr] Add python bindings for TensorExprKernel"


Differential Revision: [D27243175](https://our.internmc.facebook.com/intern/diff/D27243175)

[ghstack-poisoned]

ZolotukhinM commented on test/test_tensorexpr_pybind.py:
```diff
@@ -59,5 +66,34 @@ def test_external_calls(self):
+
+        node = None
+        graph = torch.jit.last_executed_optimized_graph()
+        node = graph.findNode("prim::TensorExprGroup", True)
```
Instead of going through the fuser pass I would rather construct a graph here directly - it also would probably be more compact too. Dependence on the fuser is unnecessary and could accidentally result in a failure for this test should the fuser somehow change its behavior.

You could find an example of how to construct a graph from a string here:
https://github.com/pytorch/pytorch/blob/2df4acd0253f32a91b9a11a4b1e7c8d392a17a35/test/test_jit.py#L1316-L1334

You could specify the necessary shape information there as well.

ZolotukhinM reviewed (commented):
Looking better! Please find my comments inline :)

huiguoo commented on torch/csrc/jit/tensorexpr/tensorexpr_init.cpp:
```diff
@@ -430,6 +431,54 @@ void initTensorExprBindings(PyObject* module) {
+                g_inputs[i]->setType(stack[i].type());
+              }
+            }
+            TensorExprKernel(graph).run(stack);
```
When preparing a stack in `run`, it needs the inputs of the graph but `TensorExprKernel` does not have any APIs for that yet. If we separate the construction of `TensorExprKernel` from `run`, we need to add new public APIs in `TensorExprKernel` c++ code to expose the graph or the inputs. 
So the options are, 
1) add `TensorExprKernel::get_inputs()` or  `TensorExprKernel::get_graph()`
2) let the user pass in the `graph` parameter again when calling `run()` which seems redundant to the user because the user has constructed a kernel with the `graph` parameter
3) like what I did - put `run()` and `fallback()` as static methods which would construct the kernel itself (no redundancy) and run/fallback it

ZolotukhinM replied:
Please add `TensorExprKernel::graph()` API to retrieve the graph for the kernel. Whatever we do we need to keep in mind the main purpose of the API we're introducing. In this case, `run` would most probably be used for benchmarking and having to compile the kernel each time we want to run it is a very bad idea for performance.

huiguoo commented on torch/csrc/jit/tensorexpr/tensorexpr_init.cpp:
```diff
@@ -430,6 +431,54 @@ void initTensorExprBindings(PyObject* module) {
+  using TSGraph = std::shared_ptr<Graph>;
+  py::class_<TensorExprKernel>(te, "TensorExprKernel")
+      .def(py::init<const TSGraph>())
+      .def_static(
```
See above :)

huiguoo committed:
Update on "[tensorexpr] Add python bindings for TensorExprKernel"


Differential Revision: [D27243175](https://our.internmc.facebook.com/intern/diff/D27243175)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.h:
```diff
@@ -35,6 +35,10 @@ class TORCH_API TensorExprKernel {
     return codegen_->getCodeText(attr);
   }
 
+  std::shared_ptr<Graph> graph() {
```
Nit: mark this method as `const`.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/tensorexpr_init.cpp:
```diff
@@ -430,6 +431,54 @@ void initTensorExprBindings(PyObject* module) {
+            return createPyObjectForStack(std::move(stack));
+          })
+      .def(
+          "get_code_gen_stmt",
```
Nit: I think we usually use `codegen` without underscore in the middle, let's stay consistent with it.

ZolotukhinM commented on test/test_tensorexpr_pybind.py:
```diff
@@ -59,5 +63,34 @@ def test_external_calls(self):
+
+        with kernel_arena_scope():
+            kernel = torch._C._te.TensorExprKernel(graph)
+            res1 = kernel.run((x, y, z))
```
Do you mind adding a test with scalar arguments as well?

huiguoo replied:
Sure, I'll add one!

ZolotukhinM reviewed (approved):
Looks good! 👍

huiguoo committed:
Update on "[tensorexpr] Add python bindings for TensorExprKernel"


Differential Revision: [D27243175](https://our.internmc.facebook.com/intern/diff/D27243175)

[ghstack-poisoned]
